Build platform-specific wheels containing libmagic

ddelange opened the pull request:
Hi @ahupp 👋

This PR builds self-contained wheels as discussed in #233. For Windows users, this renders `python-magic-bin` from @julian-r obsolete.

## pip install these wheels

1. @fpgmaas [published](https://github.com/ahupp/python-magic/pull/294#issuecomment-3327996857) the unofficial [`python-magic-standalone`](https://pypi.org/project/python-magic-standalone/) PyPI package based on this branch.
    ```sh
    pip install python-magic-standalone
    ```
3. Alternatively, pip can install these wheels from the [GitHub Release](https://github.com/ddelange/python-magic/releases/tag/0.4.28.post9) assets from my fork:
    ```sh
    # space separated list of URLs for --find-links
    export PIP_FIND_LINKS='https://github.com/ddelange/python-magic/releases/expanded_assets/0.4.28.post9'
    pip install --force-reinstall python-magic
    ```

The wheels diff:

```diff
- python-magic-0.4.27.tar.gz
+ python-magic-0.4.28.tar.gz
- python_magic-0.4.27-py2.py3-none-any.whl
+ python_magic-0.4.28-py2.py3-none-macosx_10_9_x86_64.whl
+ python_magic-0.4.28-py2.py3-none-macosx_11_0_arm64.whl
+ python_magic-0.4.28-py2.py3-none-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
+ python_magic-0.4.28-py2.py3-none-manylinux_2_17_ppc64le.manylinux2014_ppc64le.whl
+ python_magic-0.4.28-py2.py3-none-manylinux_2_17_s390x.manylinux2014_s390x.whl
+ python_magic-0.4.28-py2.py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
+ python_magic-0.4.28-py2.py3-none-musllinux_1_1_aarch64.whl
+ python_magic-0.4.28-py2.py3-none-musllinux_1_1_ppc64le.whl
+ python_magic-0.4.28-py2.py3-none-musllinux_1_1_s390x.whl
+ python_magic-0.4.28-py2.py3-none-musllinux_1_1_x86_64.whl
+ python_magic-0.4.28-py2.py3-none-win32.whl
+ python_magic-0.4.28-py2.py3-none-win_amd64.whl
```

The wheels:
- Come bundled with libmagic (`.dylib` on mac, `.so` on nix, `.ddl` on win) - no additional user action needed
  - latest libmagic (5.46)
    - [building from source](https://github.com/ahupp/python-magic/blob/688edf0385fa1c83fd8a82fffca03a459642392c/add_libmagic.sh#L5) on linux and macos with maximum backwards compatibility
    - using MSYS2 built libmagic on windows ([mingw32](https://packages.msys2.org/packages/mingw-w64-i686-file) and [mingw64](https://packages.msys2.org/packages/mingw-w64-x86_64-file))
  - skipping i686 linux wheels as they don't pass pytest ref https://github.com/ahupp/python-magic/pull/294/commits/ca4def380c04cdf48152709772a41ae2b08d3cfb
- Are forward compatible, they will install on cpython 2.7 and all current and future cpython 3.5+ distributions
- Are built for the major cpu architectures (and also for alpine linux)
  - ~Linux architectures limited by availability: https://pkgs.org/search/?q=file-libs~ now building from source on linux
  - Windows ARM64 is not included currently, but MSYS2 built libmagic is [available](https://packages.msys2.org/packages/mingw-w64-clang-aarch64-file)
- Are uploaded as GitHub Release assets, as well as upload to PyPI, whenever a GitHub Release is (pre)released.
  - :triangular_flag_on_post: @ahupp for this, **you still need to add `wheels.yml` as trusted publisher** [here](https://pypi.org/manage/project/python-magic/settings/publishing/)


## CI/CD

dists build with official `cibuildwheel` on GitHub Actions, and they build in parallel:
<img width="654" alt="image" src="https://github.com/ahupp/python-magic/assets/14880945/9c1b6bf6-09ab-471a-a44b-733beca053da">

- The PyPI publish step is failing here ^ because my fork is not a trusted publisher (see :triangular_flag_on_post: above).
- For any system not covered by these wheels, pip will fall back to the source distribution, which expects libmagic to be available at runtime.

fix #137, fix #288, fix #225, fix #276, fix #248, fix #87, fix #139, fix #233, fix #73, fix #60, fix #34, fix #293, fix #233, fix #278, fix #262, fix #248, fix #238, fix #145, fix #61, fix #12, fix #295, fix #311, fix #312, fix #313, fix #321, fix #332, fix #249, fix #333, fix #350, fix #351, fix #287, fix #335

ddelange committed:
Build platform-specific wheels containing libmagic

ddelange commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,134 @@
+      with:
+        files: dist/*
+
+    - name: Upload to PyPI
```
that's cool, thanks for sharing!

@ahupp shall I make that change and you set it up on PyPI side?

ddelange replied:
Sounds like trusted publishing is the way to go. I recently got this email:

Hi ddelange!
Earlier this year, [we announced](https://blog.pypi.org/posts/2023-05-25-securing-pypi-with-2fa/) that PyPI would require all users to enable a form of two-factor authentication on their accounts by the end of 2023.

Keeping your PyPI account secure is important to all of us. We encourage you to enable two-factor authentication on your PyPI account as soon as possible.

What forms of 2FA can I use?
We currently offer two main forms of 2FA for your account:

[Security device](https://pypi.org/help/#utfkey) including modern browsers (preferred) (e.g. Yubikey, Google Titan)
[Authentication app](https://pypi.org/help/#totp) (e.g. Google Authenticator)
Once one of these secure forms is enabled on your account, you will also need to use either [Trusted Publishers](https://docs.pypi.org/trusted-publishers/) (preferred) or [API tokens](https://pypi.org/help/#apitoken) to upload to PyPI.

What do I do if I lose my 2FA device?
As part of 2FA enrollment, you will receive one-time use recovery codes. One of them must be used to confirm receipt before 2FA is fully active. Keep these recovery codes safe - they are equivalent to your 2FA device. Should you lose access to your 2FA device, use a recovery code to log in and swap your 2FA to a new device.

Read more about[recovery codes](https://pypi.org/help/#recoverycodes).

Why is PyPI requiring 2FA?
Keeping all users of PyPI is a shared responsibility we take seriously. Strong passwords combined with 2FA is a recognized secure practice for over a decade.

We are requiring 2FA to protect your account and the packages you upload, and to protect PyPI itself from malicious actors. The most damaging attacks are account takeover and malicious package upload.

To see this and other security events for your account, visit your [account security history](https://pypi.org/manage/account/#account-events).

Read more on [this blog post.](https://blog.pypi.org/posts/2023-05-25-securing-pypi-with-2fa/#why-now)

If you run into problems, read the [FAQ page](https://pypi.org/help/). If the solutions there are unable to resolve the issue, contact us via [support@pypi.org](mailto:support@pypi.org).

Thanks,
The PyPI Admins

ddelange replied:
https://github.com/ahupp/python-magic/pull/294/commits/4a715e2b849373ddec19a2c9191090a706afeca9

ddelange replied:
@ahupp so the last thing for you to do is adding this repo as trusted publisher to https://pypi.org/manage/project/python-magic/settings/publishing/

stumpylog commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,134 @@
+name: GH
+
+permissions:
+  contents: write
```
This is for the release asset upload correct?  It could be scoped a little bit smaller to only that job

ddelange replied:
a437409

ddelange committed:
Move persmissions into job scope, remove ABI3 reference

ahupp commented:
This is huge, thank you!   Apology for the delay I thought I'd commented earlier but guess not.  I'll look this over soon; I didn't quite understand how bad the binary dep situation was expecially on windows.

ddelange committed:
Switch to PyPI trusted publishing

ddelange commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,131 @@
+        only: ${{ matrix.only }}
+      env:
+        CIBW_BUILD_VERBOSITY: 1
+        CIBW_BEFORE_BUILD: 'bash -c "make install_libmagic"'
```
```suggestion
        CIBW_BEFORE_BUILD: bash -c 'make install_libmagic'
        CIBW_TEST_COMMAND: python -c 'import magic; magic.Magic(mime=True).from_buffer("#!/usr/bin/env python\nprint(42)")'
```

ddelange replied:
1adc0a5

stumpylog commented:
I don't have anything to do with the project, so I can't help.  I just dropped by to look at something else and happened to notice some things

patacca commented:
One month has passed since last activity, I think this PR is badly needed to fix the issues with windows. Would it be possible to move things forward or at least give an estimated schedule?

ddelange committed:
Add CIBW_TEST_COMMAND and indent Makefile

ddelange committed:
Merge branch 'master' of https://github.com/ahupp/python-magic into abi3-wheels

* 'master' of https://github.com/ahupp/python-magic:
  python-magic tests: add test files for elf and json, use to test flags
  Magic.__init__: add kwargs to enable/disable different types of magic detection
  typing stubs: add Magic.__init__ extension & follow_symlinks args

ddelange committed:
Fix CI for macos

ddelange committed:
Add dependabot.yml

ddelange committed:
Migrate actions/upload-artifact@v4

ref https://github.com/actions/download-artifact/blob/v4.1.1/docs/MIGRATION.md#multiple-uploads-to-the-same-named-artifact

ddelange committed:
Ensure magic.mgc packaged in wheel gets recognised

ddelange commented:
Wheels now get installed and tested inside the cibuildwheel docker environment. I also tested the latest wheels on google colab:
```py
# pip install python-magic --force-reinstall --find-links https://github.com/ddelange/python-magic/releases/expanded_assets/0.4.28
>>> import magic
>>> magic.Magic(mime=True).from_buffer(b'\x00\x00\x00\x1cftypisom\x00\x00\x02\x00isomiso2mp41\x00')
'video/mp4'
```

> - Are uploaded as GitHub Release assets, as well as upload to PyPI, whenever a GitHub Release is (pre)released.
>   - For this, you still need to add this repo as trusted publisher to https://pypi.org/manage/project/python-magic/settings/publishing/

@ahupp one last TODO for you before/after merging ^

martin-liu commented:
@ahupp Any update?

jmoraleda commented:
I confirm that `pip install python-magic --force-reinstall --find-links https://github.com/ddelange/python-magic/releases/expanded_assets/0.4.28` works for me on MSW

On Linux, I feel that libmagic is easy enough to install from your package manager, so I am not certain we should include it by default

ddelange commented:
hi @jmoraleda :wave: please see the discussion [here](https://github.com/ahupp/python-magic/issues/233#issuecomment-1698717537)

if a package has reasons to refrain from binary distributions (bdist_wheel, .whl) e.g. for linux, it's as simple as not publishing wheels for linux. then, on linux systems, pip will fall back to the source distribution (sdist, tar.gz). i personally dont see any reason to refrain from publishing linux wheels.

ddelange commented:
if you want to circumvent the binary distribution because you want to use your own libmagic, you can add `--no-binary python-magic` to your pip install command to force a fallback to source distribution

jmoraleda commented:
Thank you again @ddelange. Definitely having MSW wheels available is a huge plus. With respect to Linux, as long as there is a mechanism for installing using the system libraries and is prominently mentioned in the installation page, I agree there is no reason from not publishing linux wheels as well.

jmoraleda commented:
Also, in debian like distributions python-magic is actually packaged by the system maintainers to already refer to the installed libraries, so if one installs from [there](https://packages.debian.org/bookworm/python3-magic) with the distribution package manager, then avoiding duplicate binaries is already achieved, so even more of a reason to include wheels for everything in the pypi distribution.

ddelange committed:
Add note about --no-binary to the installation instructions

ddelange committed:
Separate mac versions

ddelange committed:
Update cibuildwheel

ddelange committed:
Bump pypa/gh-action-pypi-publish@v1.8.14

ddelange commented:
last commit is green: https://github.com/ddelange/python-magic/actions/runs/8554996670

ddelange committed:
Bump cibuildwheel docker images

ddelange committed:
Revert "Bump cibuildwheel docker images"

This reverts commit 94718d593f9c3a602eab1ace0478ad57f5c36058.

ddelange committed:
Move magic.mgc injection into Magic class

ddelange committed:
Build on more recent cibw images

ddelange commented:
found out that the `manylinux2014` images are pretty old and hence were getting a pretty old libmagic:
```
Package file-libs-5.11-37.el7.x86_64 already installed and latest version
```
upgraded now to `manylinux_2_28` (rhel/almalinux 8):
```
Package file-libs-5.33-25.el8.x86_64 is already installed.
```

ddelange committed:
Use hls mp4 (recent libmagic only) for testing

ddelange committed:
Revert "Use hls mp4 (recent libmagic only) for testing"

This reverts commit dc075e9d8102fe49f99f5a16542ee42dde31824d.

manylinux_2_28 is not available for i686

ddelange commented:
the musllinux wheels are getting libmagic 5.45 ref https://pkgs.alpinelinux.org/package/edge/main/x86/libmagic

ddelange commented:
could consider [building from source](https://github.com/file/file/blob/FILE5_45/INSTALL#L51) in the manylinux images, to get 5.45 instead of 5.33...

ddelange commented:
@ahupp do you think that's something for a follow-up PR?

ddelange committed:
Install from source

ddelange commented:
@ahupp building from source done in fe62a26

> - Are uploaded as GitHub Release assets, as well as upload to PyPI, whenever a GitHub Release is (pre)released.
>   - For this, you still need to add this repo as trusted publisher to https://pypi.org/manage/project/python-magic/settings/publishing/

one last TODO for you before/after merging ^

ddelange commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,134 @@
+      env:
+        CIBW_BUILD_VERBOSITY: 1
+        CIBW_BEFORE_BUILD: bash "add_libmagic.sh"
+        CIBW_TEST_COMMAND: python -c "import magic; assert magic.Magic(mime=True).from_buffer(b'\x00\x00\x00\x1cftypiso5\x00\x00\x00\x01isomiso5hlsf\x00\x00') == 'video/mp4'"
```
a simple smoke test: this is an HLS MP4 video, only recognised in recent versions of libmagic.

ddelange replied:
this also runs (without PyPI upload) on pull requests, so hereby adding CI

ddelange replied:
fwiw, I tried switching from this smoketest to `pytest` (https://github.com/ddelange/python-magic/commit/84667ee7d3d6bbc6bcf2400105ed322de22417fa), but got quite many errors on Windows:

<details>
<summary>Logs</summary>

https://github.com/ddelange/python-magic/actions/runs/8966931092/job/24623426057#step:4:1293

```pytb
_________________________ MagicTest.test_descriptions _________________________
  self = <test.python_magic_test.MagicTest testMethod=test_descriptions>
      def test_descriptions(self):
          m = magic.Magic()
          os.environ["TZ"] = "UTC"  # To get last modified date of test.gz in UTC
          try:
  >           self.assert_values(
                  m,
                  {
                      "elf-NetBSD-x86_64-echo": (
                          "ELF 64-bit LSB shared object, x86-64, version 1 (SYSV)",
                          "ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /libexec/ld.elf_so, for NetBSD 8.0, not stripped",
                      ),
                      "magic._pyc_": "python 2.4 byte-compiled",
                      "test.pdf": (
                          "PDF document, version 1.2",
                          "PDF document, version 1.2, 2 pages",
                          "PDF document, version 1.2, 2 page(s)",
                      ),
                      "test.gz": (
                          'gzip compressed data, was "test", from Unix, last '
                          "modified: Sun Jun 29 01:32:52 2008",
                          'gzip compressed data, was "test", last modified'
                          ": Sun Jun 29 01:32:52 2008, from Unix",
                          'gzip compressed data, was "test", last modified'
                          ": Sun Jun 29 01:32:52 2008, from Unix, original size 15",
                          'gzip compressed data, was "test", '
                          "last modified: Sun Jun 29 01:32:52 2008, "
                          "from Unix, original size modulo 2^32 15",
                          'gzip compressed data, was "test", last modified'
                          ": Sun Jun 29 01:32:52 2008, from Unix, truncated",
                      ),
                      "text.txt": "ASCII text",
                      "test.snappy.parquet": ("Apache Parquet", "Par archive data"),
                      "test.json": "JSON text data",
                  },
                  buf_equals_file=False,
              )
  D:\a\python-magic\python-magic\test\python_magic_test.py:131: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  D:\a\python-magic\python-magic\test\python_magic_test.py:54: in assert_values
      self.assertIn(value, expected_value)
  E   AssertionError: 'ASCII text, with CRLF line terminators' not found in ('ASCII text',)
  _____________________ MagicTest.test_descriptions_no_json _____________________
  self = <test.python_magic_test.MagicTest testMethod=test_descriptions_no_json>
      def test_descriptions_no_json(self):
          m = magic.Magic(check_json=False)
          os.environ["TZ"] = "UTC"  # To get last modified date of test.gz in UTC
          try:
  >           self.assert_values(
                  m,
                  {
                      "elf-NetBSD-x86_64-echo": (
                          "ELF 64-bit LSB shared object, x86-64, version 1 (SYSV)",
                          "ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /libexec/ld.elf_so, for NetBSD 8.0, not stripped",
                      ),
                      "magic._pyc_": "python 2.4 byte-compiled",
                      "test.pdf": (
                          "PDF document, version 1.2",
                          "PDF document, version 1.2, 2 pages",
                          "PDF document, version 1.2, 2 page(s)",
                      ),
                      "test.gz": (
                          'gzip compressed data, was "test", from Unix, last '
                          "modified: Sun Jun 29 01:32:52 2008",
                          'gzip compressed data, was "test", last modified'
                          ": Sun Jun 29 01:32:52 2008, from Unix",
                          'gzip compressed data, was "test", last modified'
                          ": Sun Jun 29 01:32:52 2008, from Unix, original size 15",
                          'gzip compressed data, was "test", '
                          "last modified: Sun Jun 29 01:32:52 2008, "
                          "from Unix, original size modulo 2^32 15",
                          'gzip compressed data, was "test", last modified'
                          ": Sun Jun 29 01:32:52 2008, from Unix, truncated",
                      ),
                      "text.txt": "ASCII text",
                      "test.snappy.parquet": ("Apache Parquet", "Par archive data"),
                  },
                  buf_equals_file=False,
              )
  D:\a\python-magic\python-magic\test\python_magic_test.py:209: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  D:\a\python-magic\python-magic\test\python_magic_test.py:54: in assert_values
      self.assertIn(value, expected_value)
  E   AssertionError: 'ASCII text, with CRLF line terminators' not found in ('ASCII text',)
  _____________________ MagicTest.test_descriptions_no_soft _____________________
  self = <test.python_magic_test.MagicTest testMethod=test_descriptions_no_soft>
      def test_descriptions_no_soft(self):
          m = magic.Magic(check_soft=False)
  >       self.assert_values(
              m,
              {
                  "elf-NetBSD-x86_64-echo": (
                      "data",
                      "ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /libexec/ld.elf_so, for NetBSD 8.0, not stripped",
                  ),
                  "magic._pyc_": "data",
                  "test.pdf": "ASCII text",
                  "test.gz": "data",
                  "text.txt": "ASCII text",
                  "test.snappy.parquet": "data",
                  "test.json": "JSON text data",
              },
              buf_equals_file=False,
          )
  D:\a\python-magic\python-magic\test\python_magic_test.py:168: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  D:\a\python-magic\python-magic\test\python_magic_test.py:54: in assert_values
      self.assertIn(value, expected_value)
  E   AssertionError: 'ASCII text, with CRLF line terminators' not found in ('ASCII text',)
  __________________________ MagicTest.test_mime_types __________________________
  self = <test.python_magic_test.MagicTest testMethod=test_mime_types>
      def test_mime_types(self):
          dest = os.path.join(MagicTest.TESTDATA_DIR, b"\xce\xbb".decode("utf-8"))
          shutil.copyfile(os.path.join(MagicTest.TESTDATA_DIR, "lambda"), dest)
          try:
              m = magic.Magic(mime=True)
  >           self.assert_values(
                  m,
                  {
                      "elf-NetBSD-x86_64-echo": (
                          "application/x-pie-executable",
                          "application/x-sharedlib",
                      ),
                      "magic._pyc_": (
                          "application/octet-stream",
                          "text/x-bytecode.python",
                          "application/x-bytecode.python",
                      ),
                      "test.pdf": "application/pdf",
                      "test.gz": ("application/gzip", "application/x-gzip"),
                      "test.snappy.parquet": "application/octet-stream",
                      "text.txt": "text/plain",
                      b"\xce\xbb".decode("utf-8"): "text/plain",
                      b"\xce\xbb": "text/plain",
                      "test.json": "application/json",
                  },
                  buf_equals_file=False,
              )
  D:\a\python-magic\python-magic\test\python_magic_test.py:102: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  D:\a\python-magic\python-magic\test\python_magic_test.py:54: in assert_values
      self.assertIn(value, expected_value)
  E   AssertionError: "cannot open `D:\\\\a\\\\python-magic\\\\python-magic\\\\test\\\\testdata\\\\\u03bb' (No such file or directory)" not found in ('text/plain',)
  ___________________________ MagicTest.test_symlink ____________________________
  self = <test.python_magic_test.MagicTest testMethod=test_symlink>
      def test_symlink(self):
          # TODO: 3.0
          if not hasattr(tempfile, "TemporaryDirectory"):
              return
      
          with tempfile.TemporaryDirectory() as tmp:
              tmp_link = os.path.join(tmp, "test_link")
              tmp_broken = os.path.join(tmp, "nonexistent")
      
              os.symlink(
                  os.path.join(self.TESTDATA_DIR, "test.pdf"),
                  tmp_link,
              )
      
              os.symlink("/nonexistent", tmp_broken)
      
              m = magic.Magic()
              m_follow = magic.Magic(follow_symlinks=True)
  >           self.assertTrue(m.from_file(tmp_link).startswith("symbolic link to "))
  E           AssertionError: False is not true
  D:\a\python-magic\python-magic\test\python_magic_test.py:373: AssertionError
```

</details>

ddelange committed:
Documentation and readability

ddelange committed:
Build macos wheels with maximum backwards compatibility

ddelange committed:
Use CIBW_SKIP

i686 fails pytest with:

/project/test/python_magic_test.py:54: in assert_values
      self.assertIn(value, expected_value)
  E   AssertionError: 'gzip compressed data, was test, last modified: *Invalid datetime*, from Unix' not found in ('gzip compressed data, was test, from Unix, last modified: Sun Jun 29 01:32:52 2008', 'gzip compressed data, was test, last modified: Sun Jun 29 01:32:52 2008, from Unix', 'gzip compressed data, was test, last modified: Sun Jun 29 01:32:52 2008, from Unix, original size 15', 'gzip compressed data, was test, last modified: Sun Jun 29 01:32:52 2008, from Unix, original size modulo 2^32 15', 'gzip compressed data, was test, last modified: Sun Jun 29 01:32:52 2008, from Unix, truncated')

cclauss reviewed (approved):
# Nice!!

AbdealiLoKo commented:
Was looking for exactly something to simploify the binary installations for python-magic and came across this PR.

Looks like it is very close to the finish line ! Anything I could do to help ?

ddelange commented:
Hi @AbdealiLoKo :wave: to fetch the wheels, see the `pip install` command in the PR description. From my side this is PR ready for review.

ddelange committed:
Apply suggestions from code review

Co-authored-by: Christian Clauss <cclauss@me.com>

AbdealiLoKo commented:
So, I tried this out:

```python
$ venv/bin/pip install python-magic
Successfully installed python-magic-0.4.27

$ venv/bin/ipython
In [1]: import magic
In [2]: magic.detect_from_content(open('./README.md', 'rb').read(2048))
Out[2]: FileMagic(mime_type='text/plain', encoding='us-ascii', name='ASCII text')

$ venv/bin/pip uninstall python-magic
Successfully uninstalled python-magic-0.4.27

$ venv/bin/pip install python-magic --force-reinstall --find-links https://github.com/ddelange/python-magic/releases/expanded_assets/0.4.28.post6
Successfully installed python-magic-0.4.28

$ venv/bin/ipython
In [2]: magic.detect_from_content(open('./README.md', 'rb').read(2048))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[2], line 1
----> 1 magic.detect_from_content(open('./README.md', 'rb').read(2048))

File ~/corridor-platforms/venv/lib/python3.10/site-packages/magic/__init__.py:479, in _add_compat.<locals>.deprecation_wrapper.<locals>._(*args, **kwargs)
    473 def _(*args, **kwargs):
    474     warnings.warn(
    475         "Using compatibility mode with libmagic's python binding. "
    476         "See https://github.com/ahupp/python-magic/blob/master/COMPAT.md for details.",
    477         PendingDeprecationWarning)
--> 479     return fn(*args, **kwargs)

File ~/corridor-platforms/venv/lib/python3.10/site-packages/magic/compat.py:286, in detect_from_content(byte_content)
    280 def detect_from_content(byte_content):
    281     '''Detect mime type, encoding and file type from bytes
    282
    283     Returns a `FileMagic` namedtuple.
    284     '''
--> 286     return _create_filemagic(mime_magic.buffer(byte_content),
    287                              none_magic.buffer(byte_content))

File ~/corridor-platforms/venv/lib/python3.10/site-packages/magic/compat.py:248, in _create_filemagic(mime_detected, type_detected)
    247 def _create_filemagic(mime_detected, type_detected):
--> 248     splat = mime_detected.split('; ')
    249     mime_type = splat[0]
    250     if len(splat) == 2:

AttributeError: 'NoneType' object has no attribute 'split'
```

This error comes for all the files I have tried so far:
1. JAR file
2. YAML file
3. Excel file
4. MP4 file

ddelange commented:
can you try `from_buffer` instead of `detect_from_content`?

see also https://github.com/ahupp/python-magic/blob/0.4.27/magic/__init__.py#L427-L430 and https://github.com/ahupp/python-magic/blob/0.4.27/COMPAT.md?plain=1#L13

ddelange committed:
Merge branch 'master' of ahupp/python-magic into abi3-wheels

* 'master' of https://github.com/ahupp/python-magic:
  Update README.md
  README.md: Add a badge for GitHub Actions
  Update README.md
  Update ci.yml
  GitHub Action to replace Travis CI
  Delete .travis.yml
  Travis CI: Test on Python 3.13 beta

ddelange committed:
Fix compat.py now that bundled libmagic is preferred

ddelange committed:
Fix https://github.com/ahupp/python-magic/issues/321

ddelange committed:
Use sudo on ubuntu-latest in ci.yml

ddelange commented:
@AbdealiLoKo I've released 0.4.28.post7 containing eba05b6, and verified that the compat module now correctly loads the `magic.mgc` file bundled in the wheel.

In compat.py, `mime_magic.load()` and `none_magic.load()` were silently returning a -1, because `load_lib()` now prefers the (recent) libmagic bundled in the wheel over the (older) system libmagic, which then caused `magic.mgc` not to be found in the standard paths. The fix now points to the bundled `magic.mgc` in this case.

On Google Colab:
```pycon
# pip install python-magic --force-reinstall --find-links https://github.com/ddelange/python-magic/releases/expanded_assets/0.4.28.post7
>>> import magic
>>> magic.detect_from_content(b'\x00\x00\x00\x1cftypiso5\x00\x00\x00\x01isomiso5hlsf\x00\x00')
FileMagic(mime_type='video/mp4', encoding='binary', name='ISO Media, MP4 Base Media v5 ')
>>> magic.from_buffer(b'\x00\x00\x00\x1cftypiso5\x00\x00\x00\x01isomiso5hlsf\x00\x00')
'ISO Media, MP4 Base Media v5 '
```

Thanks for catching this!

ddelange committed:
Fix sudo not available on windows-latest

ddelange committed:
Merge branch 'ahupp:master' into abi3-wheels

cclauss commented:
What are the remaining TODOs on this pull request?

ddelange commented:
> - Are uploaded as GitHub Release assets, as well as upload to PyPI, whenever a GitHub Release is (pre)released.
>   - For this, you still need to add this repo as trusted publisher to https://pypi.org/manage/project/python-magic/settings/publishing/

- TODO for @ahupp before/after merging ^
- I should probably add to CHANGELOG

ddelange committed:
Add entries in CHANGELOG

AbdealiLoKo commented:
The compat part (which is what I generally use to avoid confusion with the non-pypi magic - works with post7 :)
Thanks ! This works well for me.

+1 to merging and releasing it !

cclauss commented:
Many people have commented on this pull request but there are very few reviews.

When you think a pull request is useful and is ready to be merged, ___please consider giving it a positive review___.

Every check mark ✔️ at the top right of this page gives project maintainers confidence that the proposed changes have been read through and deemed both useful and safe to merge into the codebase.  Lots of 👍 and "_what is the ETA?_" comments are easier for maintainers to ignore than ✔️✔️✔️✔️✔️ from several different reviewers.

Anyone can review a pull request on GitHub.  To do so here:
1. Scroll to the top of this page.
2. Click the `Files changed` tab and read through each file carefully looking for potential issues.
3. Click the `Review changes` button.
4. Click `Approve` (or one of the other options) and add comments only if they have not already been stated in the PR.
5. Click `Submit review` so that your ✔️ will be added to the list.

ddelange committed:
Merge branch 'master' of https://github.com/ahupp/python-magic into abi3-wheels

* 'master' of https://github.com/ahupp/python-magic:
  log warning on ctypes load error, adapted from #279
  smartos support, adapted from #132
  rename no_json test to avoid duplicate function definitions
  handle unknown platforms gracefully in loader.py
  Clean up loader.py

ddelange committed:
Fix test

cclauss commented on magic/loader.py:
```diff
@@ -1,60 +1,85 @@
+    """Yield possible libmagic library names on Linux."""
+    prefixes = ('libmagic.so.1', 'libmagic.so')
+
+    for i in prefixes:
```
This is not an integer (i).
```suggestion
    for prefix in prefixes:
```

ddelange replied:
f7341ce

cclauss commented on magic/loader.py:
```diff
@@ -1,60 +1,85 @@
+        "msys-magic-1",
+    )
+
+    for i in prefixes:
```
```suggestion
    for prefix in prefixes:
```

ddelange replied:
f7341ce

ddelange committed:
PR Suggestions

cclauss commented on magic/loader.py:
```diff
@@ -58,13 +58,13 @@ def _lib_candidates_windows():
+        yield os.path.join(os.path.abspath('.'), '%s.dll' % fname)
         # find_library searches in %PATH% but not the current directory
-        yield find_library(i)
+        yield find_library(fname)
```
Should we remove lines 93 and 94 below because a library can be loaded even if it is not a full path?
```python
        if not os.path.exists(lib):
            continue
```

ddelange replied:
ah youre right, the line you commented on is not a full path. @ahupp how about partially reverting https://github.com/ahupp/python-magic/commit/a9e627644efcee42b70cf7634f0ec6d2bab318c8 ?

cclauss replied:
You are correct.  My mistake.

ddelange commented:
thanks for the reviews!

ddelange commented on add_libmagic.sh:
```diff
@@ -0,0 +1,70 @@
+    # install from source
+    # https://www.darwinsys.com/file/
+    # https://github.com/file/file/blob/FILE5_45/INSTALL#L51
+    version="file-5.45"
+    (
+        tmpfile="$(mktemp)" &&
+        curl -sSLo "${tmpfile}" "https://astron.com/pub/file/${version}.tar.gz" &&
+        tar xvf "${tmpfile}" &&
+        cd "${version}" &&
+        ./configure &&
+        make &&
+        make install &&
+        make installcheck &&
+        cd .. &&
+        rm -r "${version}"
+    ) || (cd .. && rm -r "${version}" && false)
```
```suggestion
    version="file-5.45"
    (
        tmpfile="$(mktemp)" &&
        curl -sSLo "${tmpfile}" "https://astron.com/pub/file/${version}.tar.gz" &&
        tar xvf "${tmpfile}" &&
        cd "${version}" &&
        ./configure &&
        make &&
        make install &&
        make installcheck &&
        cd .. &&
        rm -r "${version}"
    )
```

ddelange committed:
Apply suggestions from code review

Co-authored-by: Frank Dana <ferdnyc@gmail.com>

ddelange commented on magic/compat.py:
```diff
@@ -16,6 +17,13 @@
 
 _libraries = {}
 _libraries['magic'] = loader.load_lib()
+magic_file = None
+if not os.environ.get("MAGIC"):
+    # wheels package the mime database in this directory
+    # prefer it when no magic file is specified by the user
+    mime_db = os.path.join(os.path.dirname(__file__), 'magic.mgc')
+    if os.path.exists(mime_db):
+        magic_file = mime_db
```
one way to avoid code duplication (and cyclic import) would be to add `from magic.compat import magic_file` on top of `__init__.py`, and use it as default argument: 
```py
Magic.__init__(magic_file=magic_file, ...)
```

option 2 would be to put the snippet in a new file and import it in both places (avoids cyclic import as well).

any preference here @ahupp?

ferdnyc replied:
> add `from magic.compat import magic_file` on top of `__init__.py`,

I don't think that's a good idea, as it means that it'd be impossible to import any part of python-magic without also importing `magic.compat`. A `compat` module, specifically, should only be imported when needed.

The code _could_ be moved to the top level of `__init__.py`, and imported into `compat.py` with `from . import magic_file`. I think that would be a better direction for things to flow, if going that route.

ferdnyc replied:
Oh, never mind. `__init__.py` already unconditionally imports `magic.compat` in its `_add_compat()` function. Which is always run on import.

ddelange committed:
Revert partially: fix install on Windows

ddelange commented:
Hi :wave: I'll be AFK until end of June. @ahupp feel free to take over my branch, or merge as is!
[![https://http.cat/301](https://http.cat/301)](https://http.cat/301)

ddelange committed:
Merge branch 'master' into abi3-wheels

Privat33r-dev commented on .github/workflows/ci.yml:
```diff
@@ -19,13 +19,10 @@ jobs:
         with:
           python-version: ${{ matrix.python-version }}
           allow-prereleases: true
+      - run: ${{ ( startsWith( matrix.os, 'windows' ) && 'bash add_libmagic.sh' ) || 'sudo -E bash add_libmagic.sh' }}
```
It's a strange command, would it be required from the user to use it on each installation?

cclauss replied:
```suggestion
      - if: runner.os == 'Windows'
        shell: bash
        run: add_libmagic.sh || sudo -E bash add_libmagic.sh
```

Privat33r-dev replied:
> Each _**Windows**_ installation.

Would it be possible to make this command a part of the setup? Isn't the main idea of this PR to make installation of binaries as a part of the PIP package?

Why would admin privileges be required? Isn't it possible to download the DLL in the same folder where the package is? If not, would it be possible to ship binaries together?

As an alternative solution, we can require package with binaries as a sub dependency.

ddelange replied:
`add_libmagic.sh`, introduced in this PR, adds the assets to the current folder which need to be packaged into the wheel. since `ci.yml` is installing from source (and not from wheel), this command needs to run here, regardless of platform (cc @cclauss)

essentially you're seeing an install from source here. install from wheel won't be needing this command. and if you already have libmagic installed yourself, you also won't be needing this command to install from source.

all in all, this shell script is only a helper for github actions, and won't be needed by PyPI users.

I hope that answers all your questions!

ddelange replied:
```suggestion
      - name: Copy libmagic into magic dir
        run: ${{ ( startsWith( matrix.os, 'windows' ) && 'bash add_libmagic.sh' ) || 'sudo -E bash add_libmagic.sh' }}
```

Privat33r-dev commented on CHANGELOG:
```diff
@@ -1,10 +1,14 @@
-Changes to 0.4.29:
+Changes to 0.4.28:
```
Why version got reduced? It might even make sense to bump major version

ddelange replied:
0.4.28 is not yet released, correcting the typo here

Privat33r-dev replied:
Thank you for the clarification, it's resolved. 

I lack permissions here to mark this conversation as resolved :)

Privat33r-dev commented on README.md:
```diff
@@ -45,27 +44,53 @@ You can also combine the flag options:
+apt-get update && apt-get install -y libmagic1
+# Alpine
+apk add --update libmagic
+# RHEL
```
It would be nice to mention fedora as well since it's more [popular](https://distrowatch.com/dwres.php?resource=popularity). Also I'd suggest to recommend dnf instead as a more modern alternative.

Privat33r-dev replied:
> is dnf installed by default? if yes, can you provide the relevant command?

DNF stands for Dandified YUM and is basically an improved version of the YUM package manager.

[Red Hat Doc](https://docs.redhat.com/it/documentation/red_hat_enterprise_linux/9/html-single/managing_software_with_the_dnf_tool/index#proc_providing-feedback-on-red-hat-documentation_managing-software-with-the-dnf-tool)
>In Red Hat Enterprise Linux (RHEL) 9, use the DNF utility to manage software. For compatibility reasons with previous major RHEL versions, you can still use the yum command. However, in RHEL 9, yum is an alias for dnf which provides a certain level of compatibility with yum.

[yum-dnf manual from opensource.com](https://opensource.com/article/18/8/guide-yum-dnf)
> [about DNF] It was first introduced in Fedora 18 in a testable state (i.e., tech preview), but it's been Fedora's default package manager since Fedora 22.

The command is the same:
`dnf install SAME_PACKAGE_AS_FOR_YUM`

Privat33r-dev commented on README.md:
````diff
@@ -45,27 +44,53 @@ You can also combine the flag options:
+Depending on your system and CPU architecture, there might be no compatible wheel uploaded.
+However, precompiled libmagic might still be available for your system:
+
+```sh
````
It may be beneficial to add a library installation guide for SUSE as well

ddelange replied:
can you provide the relevant command?

fwiw, I think mostly all linux flavours will be covered by the wheels in the PR description, so those users won't be needing the install from source instructions provided here.

Privat33r-dev replied:
I guess that it would be 
`zypper install file-devel`
it might as well be required to do (in vivo test is required though)
`zypper install file-magic`

Currently I don't have OpenSUSE at my disposal for tests and it's likely that it would be a default package.

Privat33r-dev reviewed (approved):
Generally LGTM apart from some minor issues. Ideally, to have better control over dependencies, I would keep binaries inside of the repo, like pyexiv2 and some other libraries do.

Privat33r-dev commented on add_libmagic.sh:
```diff
@@ -0,0 +1,71 @@
+        tar xvf "${tmpfile}" &&
+        cd "${version}" &&
+        ./configure &&
+        make &&
```
Developer's dependencies to use "make" and similar commands are not shipped by default in some distros, especially in those that are usually used by Docker. I'd suggest to ensure that necessary dependencies are installed first.

ddelange replied:
I wished `set -euxo pipefail` also effects commands inside function definitions. then we could remove `&&` from every line here (and the subshell parentheses here and the `||` on L21), and we could test `which make` on top of this function or similar.

now, it simply falls back to `install_precompiled` (L69) when curl or make or compilation fails (error will be in output).

if make is not available, in the current setup we will only have done an unnecessary curl. not the worst but could be avoided.

any ideas?

Privat33r-dev replied:
> I wished `set -euxo pipefail` also effects commands inside function definitions. then we could remove `&&` from every line here (and the subshell parentheses here and the `||` on L21), and we could test `which make` on top of this function or similar.
> 
> any ideas?

Just use it [inside](https://unix.stackexchange.com/questions/405382/set-e-inside-a-bash-function) of a function.

You made me think about other topic when you mentioned that `curl` can be executed. There is no cleanup on any step, it might be logical to make a ZIP file cleanup from external function and do it regardless of the outcome. Same for windows. Pseudocode:
install_stuff(){
 real_install()
 rm potential_zip
}

ddelange commented on README.md:
```diff
@@ -45,27 +44,53 @@ You can also combine the flag options:
+# Alpine
+apk add --update libmagic
+# RHEL
+yum install file-libs
```
```suggestion
dnf install file-libs
```

ddelange committed:
Apply suggestions from code review

Privat33r-dev commented on add_libmagic.sh:
```diff
@@ -0,0 +1,71 @@
+    (
+        version="file-5.45" &&
+        tmpfile="$(mktemp)" &&
+        curl -sSLo "${tmpfile}" "https://astron.com/pub/file/${version}.tar.gz" &&
```
Tried it on ubuntu's Docker, fails here :)
`bash: curl: command not found`

ddelange replied:
yes, the script is only intended for github actions default runner images and the cibuildwheel docker images (both have curl and make)

cclauss replied:
`sudo apt update && sudo apt install curl`

cclauss commented:
There is wonderful work in this pull request and it has four positive reviews.  Unfortunately, it has been open for ~9 months without landing.  Perhaps it would be useful to break it into three _separate_ PRs that are easier to review and land.

One PR that deals only with macOS and another that deals only with Windows might be easier to land.  Once that is done then this PR could be rebased to deal only with Linux and friends.  I know it is extra work but I sense that new momentum is needed.

estarfoo commented:
Hi! Great PR, is there any real contention about it beyond the scope of OS/distro support in `add_libmagic.sh` and  `README.md`?

Most users will only ever want the wheels from this repo. In particular, this looks like it will solidly cover usage in Docker images. Anyone who wants to use the source version and provide libmagic themselves, probably knows best how to do the latter in their environment. given info on where this package will look for the library. (Those who package python-magic for their Linux distro of choice will already have a preferred way of ensuring libmagic presence. This will probably not exactly match anything suggested in python-magic docs.)

Even for those particularly invested in having a range of setup instructions, the PR in its current state should look like a clear improvement on master, and further improvements in that are will come more easily as separate PRs (because they won’t be tied to CI scripts).

So: how about merging this without completing libmagic setup instructions for every possible platform? Seems like it already does what the PR title says.

Privat33r-dev commented:
> So: how about merging this without completing libmagic setup instructions for every possible platform? Seems like it already does what the PR title says.

Totally for it. My suggestions just show the way to improve it, but I would merge it "as is" since it already provides a huge value. "Perfect is the enemy of good".

@ahupp hopefully you can find some time to review this most discussed PR in the `python-magic`'s history :)

estarfoo commented on add_libmagic.sh:
```diff
@@ -0,0 +1,71 @@
+    # https://github.com/file/file/blob/FILE5_45/INSTALL#L51
+    (
+        version="file-5.45" &&
+        tmpfile="$(mktemp)" &&
+        curl -sSLo "${tmpfile}" "https://astron.com/pub/file/${version}.tar.gz" &&
+        tar xvf "${tmpfile}" &&
```
Just nitpicking, no blocking issues:
- `-f` (`--fail`) would ensure `curl` fails if the transfer is successful but the server responds with a non-2xx (with `-L`, also non-3xx) status code. In that case, the response probably is not usable anyway.
- This request currently returns the file directly, without redirect. To calm the paranoid, maybe omit `-L`?
- Unless the download is extremely fast and downloaded content is huge, and `curl` risks running OOM waiting for `tar` to unpack the archive (none of this is likely here), the `$tmpfile` handling can be replaced by a pipe. Assuming, of course, that `-o pipefail` remains set.

```suggestion
        curl -fsS "https://astron.com/pub/file/${version}.tar.gz" | \
        tar xv &&
```

ferdnyc replied:
@estarfoo 

Not a big deal, but if going with the piped `curl` output, personally I'd pair that with an explicit `tar xvf -`. Again, for the paranoid.

estarfoo commented on add_libmagic.sh:
```diff
@@ -0,0 +1,71 @@
+import platform, sysconfig, io, zipfile, urllib.request
+assert platform.system() == "Windows"
+machine = "x86" if sysconfig.get_platform() == "win32" else "x64"
+url = f"https://github.com/julian-r/file-windows/releases/download/v5.44/file_5.44-build104-vs2022-{machine}.zip"
```
In case anyone else wondered, 5.44 is the latest release available there, as opposed to 5.45 for the source tarball in `install_source()`.

ddelange commented:
@ahupp this also fixes failing CI on master (looks like the github actions linux runner image no longer ships libmagic by default)

ashtonpaul commented:
@ahupp @ddelange Any update on this PR and the release? 👀 Is there anything that needs to be tested or are we waiting on anything other than a review?

ddelange commented:
@ahupp please review, merge, add trusted publisher (see PR description) and release :pray:

wilhelmklopp commented:
Also would love to see this merged, it's causing [some issues for windows users](https://github.com/kolofordjango/kolo/issues/95) of our package (which depends on [mocket](https://github.com/mindflayer/python-mocket) which depends on python-magic). Thank you! 😄

cclauss commented:
@ashtonpaul @ddelange @wilhelmklopp https://github.com/ahupp/python-magic/pull/294#issuecomment-2129493694

ferdnyc commented:
@ahupp Speaking personally, this is a big, complex PR that I don't really feel qualified to _approve_ given the scope and nature of all the changes here. I guess I'm in agreement with your https://github.com/ahupp/python-magic/pull/294#issuecomment-2191597272 

> Perhaps it would be useful to break it into three _separate_ PRs that are easier to review and land.
> 
> One PR that deals only with macOS and another that deals only with Windows might be easier to land. Once that is done then this PR could be rebased to deal only with Linux and friends. I know it is extra work but I sense that new momentum is needed.

OTOH, the PR has received a number of approvals as-is, so there's ample indication that others are satisfied with it in its current form. I left some comments, which were addressed, so my input certainly shouldn't be viewed as *disapproval* by any means.

C0rn3j commented:
@ahupp This PR has 6 approvals, and from the looks of it, 0 controversy, would it be possible to review & pull it sometime soon?  
I just ran into one of the issues this fixes myself.

![image](https://github.com/user-attachments/assets/e3840b9c-1604-4e97-a48f-a9de781e7301)

nightlark commented:
I'm trying to understand what the current situation is with using libmagic for file type identification in Python, and I've found at least 13 separate packages on PyPI that are in various states of not being maintained -- of them, this PR seems to be the version that includes the most complete set of wheels with pre-compiled copies of libmagic (aside from not being available on PyPI).

Would it make sense to create a fork of python-magic based on this PR by @ddelange, perhaps under an org with at least 2 maintainers that can review PRs and publish updated wheels to help minimize the burden for a single person and hopefully avoid adding to the unmaintained package count?

@ahupp has done an great job with python-magic and this is in no way a criticism of their work or minimizing the time they've spent -- of the packages I found on PyPI, python-magic seems to be in the best overall shape. Life happens and situations change so time isn't as plentifully as it once was, and I know how it is for things to get really busy and maintaining an open source project can be draining, leading to burnout. This happens literally all the time (as evidenced by all of the libmagic PyPI packages that haven't been updated since before 2019).

ddelange commented:
@ahupp fwiw I'm happy to become a maintainer of this project 👍

stumpylog commented:
I would also be happy to help maintain.  We use this in paperless-ngx, so I'm happy to work on keeping it up to date, resolving issues, etc

ddelange committed:
Merge branch 'master' of https://github.com/ahupp/python-magic into abi3-wheels

* 'master' of https://github.com/ahupp/python-magic:
  update magic/compat.py
  simplify tests into something more delarative

ddelange committed:
Fix tests

ddelange committed:
Use older typing syntax

ddelange committed:
Use older typing syntax

ddelange commented:
Hi @ahupp 👋

I fixed the broken tests I just pulled from master. Did you forget to track the `test.py` file by any chance? I commented out that test for now.

I also upgraded this PR to build wheels and run CI using the newly released `file-5.46`.

Any response from you on this PR would be greatly appreciated 🙏

cclauss commented on .github/workflows/ci.yml:
```diff
@@ -19,13 +19,11 @@ jobs:
           python-version: ${{ matrix.python-version }}
           allow-prereleases: true
+      - name: Copy libmagic into magic dir
+        run: ${{ ( startsWith( matrix.os, 'windows' ) && 'bash add_libmagic.sh' ) || 'sudo -E bash add_libmagic.sh' }}
```
```suggestion
        run: ${{ (runner.os == 'Windows' && 'bash add_libmagic.sh')  || 'sudo -E bash add_libmagic.sh' }}
```
[GitHub Actions: Detecting the operating system](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#detecting-the-operating-system)

cclauss commented on .github/workflows/wheels.yml:
```diff
@@ -0,0 +1,141 @@
+      with:
+        python-version: '3.x'
+
+    - run: sudo apt-get install -y libmagic1
```
This will tell you that it is already installed on both Intel and ARM.

ddelange replied:
Right. But do we want to rely on github bundling libmagic in their ubuntu-latest runner image?

cclauss commented on .github/workflows/wheels.yml:
```diff
@@ -0,0 +1,141 @@
+    - name: Build source distribution
+      run: |
+        pip install --upgrade setuptools wheel pip
+        python setup.py sdist
```
This is deprecated.  Please use `build` instead.

ddelange replied:
https://github.com/ahupp/python-magic/pull/294/commits/d3e886c1395422a24ebde5069297359dc8f44e3a

ddelange commented on .github/workflows/wheels.yml:
```diff
@@ -0,0 +1,141 @@
+
+    - name: Build source distribution
+      run: |
+        pip install --upgrade setuptools wheel pip
+        python setup.py sdist
```
```suggestion
        pip install --upgrade setuptools wheel pip build
        python -m build --sdist
```

ddelange committed:
Apply suggestions from code review

Co-authored-by: Christian Clauss <cclauss@me.com>

ddelange committed:
Merge branch 'master' of https://github.com/ahupp/python-magic into abi3-wheels

* 'master' of https://github.com/ahupp/python-magic:
  add support for python 3.13
  Unbreak various things

ddelange committed:
Merge branch 'master' of https://github.com/ahupp/python-magic into abi3-wheels

* 'master' of https://github.com/ahupp/python-magic:
  format with ruff

ddelange committed:
Run ruff format

ddelange committed:
Skip install_source on windows

ddelange committed:
Typo

ddelange commented:
so I pulled master again and now Windows wheels are [failing](https://github.com/ahupp/python-magic/actions/runs/13875629725/job/38827548755?pr=294) since last week with the following error. I don't know why it now started failing. I don't see changes to https://github.com/julian-r/file-windows, nor relevant changes in `master`. Anyway, better to get rid of that dependency so I pushed for a more official libmagic wheel build for Windows. Fix incoming, with latest libmagic 4.46 as a bonus! 🎉 

```pytb
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-l6641oui\cp312-win_amd64\venv-test\Lib\site-packages\magic\__init__.py", line 120, in __init__
      magic_load(self.cookie, magic_file)
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-l6641oui\cp312-win_amd64\venv-test\Lib\site-packages\magic\__init__.py", line 384, in magic_load
      return _magic_load(cookie, coerce_filename(filename))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-l6641oui\cp312-win_amd64\venv-test\Lib\site-packages\magic\__init__.py", line 275, in errorcheck_negative_one
      raise MagicException(err)
  magic.MagicException: b"File 5.44 supports only version 18 magic files. `C:\\Users\\runneradmin\\AppData\\Local\\Temp\\cibw-run-l6641oui\\cp312-win_amd64\\venv-test\\Lib\\site-packages\\magic\\magic.mgc' is version 20"
```

ddelange committed:
Get latest libmagic from MSYS2 for Windows (#10)

ddelange commented:
for anyone interested in latest libmagic on windows: I created a new [GitHub release](https://github.com/ddelange/python-magic/releases/tag/0.4.28.post9) on my fork, and updated the pip install instructions in this PR description.

ddelange committed:
Remove dev leftover

DeflateAwning commented:
Can this please be merged? Is this library actively maintained? Seems like a serious improvement.

ddelange committed:
Merge branch 'master' into abi3-wheels

ddelange committed:
Update README.md

ddelange committed:
Mention MSYS2 for Windows in README.md

cclauss commented on .github/workflows/ci.yml:
```diff
@@ -19,13 +22,11 @@ jobs:
           python-version: ${{ matrix.python-version }}
           allow-prereleases: true
+      - name: Copy libmagic into magic dir
+        run: ${{ (runner.os == 'Windows' && 'bash add_libmagic.sh')  || 'sudo -E bash add_libmagic.sh' }}
```
On non-Windows platforms, `sudo -E bash add_libmagic.sh` will be executed.  Brew does not want / to be run under `sudo`.

ddelange replied:
the script installs from source on nix, which requires sudo. only when source install fails will it fall back to package managers (brew, apt, etc). i think i tried moving the sudo call inside the script at some point but failed

cclauss reviewed (changes requested):
It would be good to get this landed.

ddelange committed:
Merge branch 'ahupp:master' into abi3-wheels

fpgmaas commented:
Any updates here @ahupp ? I think many users are eagerly waiting for this to be available on PyPI. Even having a prerelease available on PyPI would be great!

fpgmaas commented:
Since this PR has been open for quite a while and there are no indications of it being merged anytime soon, I took the liberty of uploading the wheels to PyPI under the name `python-magic-standalone`, see https://pypi.org/project/python-magic-standalone/. It's sole purpose is to make it possible for users to install @ddelange's work directly from PyPI. I don't intend to maintain this fork, and I added a clear disclaimer to the `README`. It can be installed with any of the following commands, depending on the tool you use;

```
# pip
pip install python-magic-standalone

# uv
uv add python-magic-standalone

# poetry
poetry add python-magic-standalone
```

cclauss commented:
@fpgmaas, please consider building wheels for Python 3.14, as it is less than two weeks away.
* #355

ddelange commented:
this branch / these wheels are forward compatible, no need for changes in the build 👍

ddelange commented:
Updated the PR description. Thanks @fpgmaas :tada:

ddelange committed:
Bump macos-13 to macos-15-intel

https://github.blog/changelog/2025-09-19-github-actions-macos-13-runner-image-is-closing-down/

cclauss commented on .github/workflows/wheels.yml:
```diff
@@ -62,8 +62,8 @@ jobs:
-              && cibuildwheel --print-build-identifiers --platform macos --arch x86_64 |  jq -nRc '{"only": inputs, "os": "macos-13"}' \
-              && cibuildwheel --print-build-identifiers --platform macos --arch arm64 |  jq -nRc '{"only": inputs, "os": "macos-14"}' \
+              && cibuildwheel --print-build-identifiers --platform macos --arch x86_64 |  jq -nRc '{"only": inputs, "os": "macos-15-intel"}' \
+              && cibuildwheel --print-build-identifiers --platform macos --arch arm64 |  jq -nRc '{"only": inputs, "os": "macos-15"}' \
```
macos-26?

The currently available GitHub Actions macOS runners are:
| macOS Version | runner.arch |
|---------------|-------------|
| macos-13 | X64 |
| macos-14 | ARM64 |
| macos-15 | ARM64 |
| macos-15-intel | X64 |
| macos-26 | ARM64 |
| macos-latest | ARM64 |

ddelange replied:
this commit goes to macos-15-intel and macos-15 according to https://github.blog/changelog/2025-09-19-github-actions-macos-13-runner-image-is-closing-down/

ddelange committed:
Build wheels on cp313

CodeXex123 reviewed (approved):
Looks good

thc1102 committed:
update coerce_filename

ddelange commented on magic/__init__.py:
49982d2 is a `git cherry-pick` from https://github.com/thc1102/python-magic to fix #287 / #335

ddelange committed:
Revert "Build wheels on cp313"

This reverts commit b4c04a3c0843aef12df9414642b0d9bdb9de41c2.

ddelange committed:
Merge branch 'master' into abi3-wheels

cm253 commented:
Is this PR still waiting for something? :)

fpgmaas commented:
Hi @ahupp, just checking in to see if there are any updates on this PR. It seems quite some users are waiting for this to be merged. Are you still trying to determine if this should be merged or not, or are you simply lacking the time to properly review? I am aware that maintaining a project can be a bit of a timesink.

If time is the constraint, is there anything the community can do to help out? Is it maybe worth looking for maintainers for this project to help you manage it?

ddelange committed:
Merge branch 'ahupp:master' into abi3-wheels

ddelange committed:
Bump to file-5.47

ddelange committed:
Revert "Bump to file-5.47" because of #362

This reverts commit a530c13dcff62a0142ceded58bb9be0d32b2772c.

yvesll commented:
Hi @ahupp

Adding another real-world case that's blocked on this: the https://github.com/zephyrproject-rtos/zephyr project uses python-magic in its compliance tooling [check_compliance.py](https://github.com/zephyrproject-rtos/zephyr/blob/acc53653b4707aa40c22aec8c33f1d733e5e670f/scripts/ci/check_compliance.py#L25). 
Zephyr's requirements already handle the platform split correctly:
```bash
python-magic-bin; sys_platform == "win32"
python-magic; sys_platform != "win32"
```
This worked fine until `reuse>=6.0.0` was introduced as a base dependency: reuse now pulls in python-magic as a transitive dependency on all platforms, which conflicts with `python-magic-bin` on Windows and causes `ImportError: failed to find libmagic`

Here's a zephyr side workaround https://github.com/zephyrproject-rtos/zephyr/pull/107249 that patches the import at runtime, but it's exactly the kind of hack that shouldn't be necessary once this PR lands.

The implementation here looks solid and all checks are green. Is there anything blocking the merge from your side?

ddelange committed:
Merge branch 'master' of https://github.com/ahupp/python-magic into abi3-wheels
